toolchain/check/diagnostic_emitter.cpp

```diff
@@ -55,21 +56,33 @@ auto DiagnosticEmitter::ConvertLocImpl(SemIR::LocId loc_id, bool is_token_only,
     // TODO: Include the name of the imported library in the diagnostic.
     auto diag_loc =
         ConvertLocInFile(absolute_node_id, is_token_only, context_fn);
-    CARBON_DIAGNOSTIC(InImport, LocationInfo, "in import");
-    context_fn(diag_loc.loc, InImport);
+    InImport(diag_loc.loc, context_fn);
   }
 
   return ConvertLocInFile(final_node_id, is_token_only, context_fn);
 }
 
+// Recursion is allowed since in C++ locations, we also convert the location of
+// the C++ import.
+// NOLINTNEXTLINE(misc-no-recursion)
```

Comment:
We're generally trying to avoid recursion, and it's nice to let static analysis handle that instead of requiring inspection.

The idea here is that the call to `ConvertLocInFile` is always going to go down the other path, right? Perhaps recursion could be avoided with a lambda, e.g.:

```
  auto convert_node_id = [&](SemIR::CheckIRId check_ir_id, Parse::NodeId node_id, bool token_only) {
    const auto& tree_and_subtrees =
        tree_and_subtrees_getters_[check_ir_id.index]();
    return tree_and_subtrees.NodeToDiagnosticLoc(node_id, token_only);
  }

  ...

  InImport(convert_node_id(convert_node_id(sem_ir_->check_ir_id(), import_node_id, token_only).loc);

  ...

  convert_node_id(aboslute_node_id.check_ir_id(), absolute_node_id.node_id());
```